const-eval error: always say in which item the error occurred

RalfJung opened the pull request:
I don't see why "is this generic" should make a difference. It may be reasonable to key this on whether the error occurs in a `const fn` that was invoked by a const (making it non-obvious which constant it is) vs inside the body of the const.

r? @oli-obk

RalfJung commented on tests/ui/const-generics/defaults/default-param-wf-concrete.next.stderr:
```diff
@@ -2,7 +2,7 @@ error[E0080]: attempt to compute `u8::MAX + 1_u8`, which would overflow
    |
 LL | struct Foo<const N: u8 = { 255 + 1 }>;
-   |                            ^^^^^^^ evaluation of constant value failed here
+   |                            ^^^^^^^ evaluation of `Foo::{constant#0}` failed here
```
This isn't great, ideally it'd say something like "evaluation of unnamed constant inside `Foo` failed here" or so...

oli-obk replied:
yea, this may be sth we want to improve for all mentions of unnamed consts. The disambiguators don't help anyone but compiler devs.

RalfJung commented on tests/ui/const-generics/issues/issue-100313.stderr:
```diff
@@ -2,7 +2,7 @@ error[E0080]: writing to ALLOC0 which is read-only
    |
 LL |     x.set_false();
-   |     ^^^^^^^^^^^^^ evaluation of constant value failed here
+   |     ^^^^^^^^^^^^^ evaluation of `T::<&true>::set_false` failed here
```
Why does this use the name of the function rather than the name of the surrounding constant...?

RalfJung commented on tests/ui/consts/const-eval/const_fn_ptr_fail2.stderr:
```diff
@@ -14,7 +14,7 @@ error[E0080]: calling non-const function `double`
    |
 LL | const Z: usize = bar(double, 2); // FIXME: should fail to typeck someday
-   |                  ^^^^^^^^^^^^^^ evaluation of constant value failed here
+   |                  ^^^^^^^^^^^^^^ evaluation of `bar` failed here
```
Yeah something is very odd here: https://github.com/rust-lang/rust/issues/142010

oli-obk commented on tests/ui/borrowck/issue-81899.stderr:
```diff
@@ -2,7 +2,7 @@ error[E0080]: evaluation panicked: explicit panic
    |
 LL | const _CONST: &[u8] = &f(&[], |_| {});
-   |                        ^^^^^^^^^^^^^^ evaluation of constant value failed here
+   |                        ^^^^^^^^^^^^^^ evaluation of `f::<{closure@$DIR/issue-81899.rs:6:31: 6:34}>` failed here
```
Nope, we don't. huh. This is very weird. Even if some promotion thing were going on, we should be reporting the promoted in that case

RalfJung replied:
Right, you're sharing my confusion now. ;)

I can only look at this more closely next weekend -- I thought this would be a largely editorial change.^^

RalfJung commented:
Once https://github.com/rust-lang/rust/pull/142015 lands, do we even still want this? The first label always points inside the "root constant" (I forgot about that), so we don't have to repeat the name of that constant.

It's kind of unnecessary that we spell out whether it's a static or const but :shrug:

RalfJung committed:
const-eval error: always say in which item the error occurred

also adjust the wording a little so that we don't say "the error occurred here" for two different spans

RalfJung commented:
Okay, now this looks as expected. :) I also replaced "failed here" by "failed inside this call" when the label is pointing at the top of an (inverted) stack trace rather than the fine-grained failure location.

@rustbot ready

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15506550747/job/43662162753) [(plain)](https://github.com/rust-lang/rust/commit/c78beb81f46f8e495fdb63f29d5778458d92e843/checks/43662162753/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/fail/function_calls/exported_symbol_bad_unwind2.rs (revision `definition`) ... ok
tests/fail/function_calls/exported_symbol_bad_unwind2.rs (revision `both`) ... ok

FAILED TEST: tests/fail/const-ub-checks.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-6S9bO9" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/fail" "tests/fail/const-ub-checks.rs" "--edition" "2021"

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/fail/const-ub-checks.stderr` to the actual output
--- tests/fail/const-ub-checks.stderr
+++ <stderr output>
---
   |
LL |     let _x = UNALIGNED_READ;
   |              ^^^^^^^^^^^^^^
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.

---
   |
LL |     let _x = UNALIGNED_READ;
   |              ^^^^^^^^^^^^^^
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.

full stdout:



FAILED TEST: tests/fail/erroneous_const2.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-6S9bO9" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/fail" "tests/fail/erroneous_const2.rs" "--edition" "2021"

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/fail/erroneous_const2.stderr` to the actual output
--- tests/fail/erroneous_const2.stderr
+++ <stderr output>
---
   |
LL |     println!("{}", FOO);
   |                    ^^^
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

note: erroneous constant encountered
  --> tests/fail/erroneous_const2.rs:8:20
   |
LL |     println!("{}", FOO);
---
   |
LL |     println!("{}", FOO);
   |                    ^^^
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

note: erroneous constant encountered
  --> tests/fail/erroneous_const2.rs:8:20
   |
LL |     println!("{}", FOO);
---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.29.2/src/lib.rs:369

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-f58affc77002becb` (exit status: 1)
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:05:06
  local time: Sat Jun  7 10:10:51 UTC 2025
  network time: Sat, 07 Jun 2025 10:10:51 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15507019898/job/43663160655) [(plain)](https://github.com/rust-lang/rust/commit/03c440b69ec897cab268b4daeaf7dbfdfc6a6bd3/checks/43663160655/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
diff of stderr:

2   --> $DIR/const-evalutation-ice.rs:10:22
3    |
4 LL | pub const N: usize = 0 - (mem::size_of::<S>() != 400) as usize;
-    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of constant value failed here
+    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `N` failed here
6 
7 error: aborting due to 1 previous error
8 


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args const-evalutation-ice.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" "/checkout/tests/rustdoc-ui/const-evalutation-ice.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-o" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/rustdoc-ui/const-evalutation-ice" "-A" "internal_features" "-Cdebuginfo=0"
stdout: none
--- stderr -------------------------------
error[E0080]: attempt to compute `0_usize - 1_usize`, which would overflow
##[error]  --> /checkout/tests/rustdoc-ui/const-evalutation-ice.rs:10:22
   |
LL | pub const N: usize = 0 - (mem::size_of::<S>() != 400) as usize;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `N` failed here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
------------------------------------------
```

</details>

oli-obk commented:
@bors r+

RalfJung commented on compiler/rustc_const_eval/src/const_eval/eval_queries.rs:
```diff
@@ -430,27 +430,15 @@ fn report_eval_error<'tcx>(
-            ("const", String::new())
-        }
-    };
+    let instance = with_no_trimmed_paths!(cid.instance.to_string());
```
Possible? Probably. But it might be very hard...

Please file an issue.

estebank replied:
Do note that structured errors can display anything that implements `IntoDiagArg` as text, so a structured error that has a `Ty<'_>` field then it gets evaluated by the long type trimming logic, so we get that behavior for free when using structured errors.
